src/ui/app/components/transactionBuilder.jsx

```diff
@@ -131,6 +135,31 @@ const TransactionBuilder = React.forwardRef(({ onConfirm }, ref) => {
         }));
       }
     },
+    async initUndelegate(account, delegation) {
+      setData({
+        fee: '',
+        stakeRegistration: '',
+        rewards: '',
+        ready: false,
+        error: '',
+      });
+      undelegateRef.current.openModal(account.index);
+      const protocolParameters = await initTx();
+      try {
+        const tx = await undelegateTx(account, delegation, protocolParameters);
+        setData({
+          tx,
+          account,
+          fee: tx.body().fee().to_str(),
+          ready: true,
+        });
+      } catch (e) {
```

Comment:
What errors are possible to receive here? Would it be possible to know whether the balance is too low in advance?

Reply:
There are a few errors possible, like transaction too big, insufficient balance, change amount to small to cover min utxo value etc. We don't handle these cases seperately, but just say Transaction not possible if the catch block is reached.
I think it's better to build a transaction first to check if the balance is too low. A lot of parameters play a role here.

Reply:
👀 sounds exciting 😆 like: thorough error handling seems possible.

> I think it's better to build a transaction first to check if the balance is too low. A lot of parameters play a role here.

Does this mean: “There could be error handling improvements in this PR”? or “what we have here is good for now”? I’m fine with either.

Reply:
Actually both of theses statements are true.
Error handling could probably be improved. I added that new feature in a lazy way and didn't looked to improve what is already working. Actually transaction are built through the serialization-library maintained by Emurgo.
I can check if the exceptions thrown by the SL for a failed transaction build can be worked with, but there is more to it.

Like Alessandro said, there is a lot of moving part when dealing with transaction and we can't know theses part upfront. (Every added UTxO cause a ripple effect changing the Tx requirements). The UTxO are provided by the coinselection algorithm. The coin-selection still a little dumb and doesn't know about the transaction itself, so it provides enough UTxO to fulfill the request (It actually might provide more to try regroup small UTxO). For this reason, when dealing with unordinary UTxO set, the transaction can fail for the sole purpose that the combination of UTxO bust some limit.

The wallet actually attempts to build the transaction 5 time before giving up and displaying "Transaction impossible".

Example: A transaction could fail the first time because it is to large (to many utxo included to fulfill the request), the second run could fail cause UTxO holding to many native assets have been included.
In the end, the wallet can hold enough ada & assets, but because of the arrangement of it, it can be impossible to send a specific transaction without making first a transaction to reorganize the UTxO in a different manners. So it is complex to know what error to display to the user anyway. Running it 1 more time could draw the lucky combination.

The wallet actually try to improve the UTxO arrangement on each transaction a user do, but there is still improvement to be done, notably on the native assets side.

For all of the above, the statement “what we have here is good for now” is also true. The design of some part of Nami can look weird at first, but there is a story explaining every decision we made. The speed to market made the success of the wallet but sure there was some compromise to be done..!

To improve the code we will need to do a feature freeze and move very carefully..!

Reply:
Ok ok !! that sounds good. That was super interesting to read. Thank you!